cns/restserver/ipam.go

```diff
@@ -891,3 +893,170 @@ func validateDesiredIPAddresses(desiredIPs []string) error {
+		return nil, ErrStoreEmpty
+	}
+
+	if service.Options[common.OptManageEndpointState] == true {
+		err := service.EndpointStateStore.Read(EndpointStoreKey, &service.EndpointState)
+		if err != nil {
+
+			if errors.Is(err, store.ErrKeyNotFound) {
+				// Nothing to retrieve.
+				logger.Printf("[GetEndpointState]  No endpoint state to retrieve.\n")
+			} else {
+				logger.Errorf("[GetEndpointState]  Failed to retrieve state, err:%v", err)
+			}
+			return nil, errors.Wrap(err, "[GetEndpointState]  Failed to retrieve state")
+		}
+		if endpointInfo, ok := service.EndpointState[endpointID]; ok {
+			logger.Warnf("[GetEndpointState] Found existing endpoint state for container %s", endpointID)
+			return endpointInfo, nil
+		}
+		return nil, errors.New("endpoint could not be found in the statefile")
+
+	}
+	return nil, ErrOptManageEndpointState
+}
+
+// UpdateEndpointHandler handles the incoming UpdateEndpoint requests with http Patch method
+func (service *HTTPRestService) UpdateEndpointHandler(w http.ResponseWriter, r *http.Request) {
+	logger.Printf("[Azure CNS] updateEndpoint")
+
+	var req cns.EndpointRequest
+	err := service.Listener.Decode(w, r, &req)
+
+	logger.Request(service.Name, &req, err)
+	if err != nil {
+		response := cns.UpdateEndpointResponse{
+			Response: cns.Response{
+				ReturnCode: types.InvalidRequest,
+				Message:    fmt.Sprintf("[Azure CNS] updateEndpoint failed with error: %s", err.Error()),
+			},
+		}
+		err = service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+		return
+	}
+
+	err = service.UpdateEndpointHelper(req)
+	if err != nil {
+		response := cns.UpdateEndpointResponse{
+			Response: cns.Response{
+				ReturnCode: types.UnexpectedError,
+				Message:    fmt.Sprintf("[Azure CNS] updateEndpoint failed with error: %s", err.Error()),
+			},
+		}
+		err = service.Listener.Encode(w, &response)
+		logger.Response(service.Name, response, response.Response.ReturnCode, err)
+	}
+	response := cns.UpdateEndpointResponse{
+		Response: cns.Response{
+			ReturnCode: types.Success,
+			Message:    "[Azure CNS] updateEndpoint retruned successfully",
+		},
+	}
+	err = service.Listener.Encode(w, &response)
+	logger.Response(service.Name, response, response.Response.ReturnCode, err)
+}
+
+// UpdateEndpointHelper updates the state of the given endpointId with HNSId or VethName
+func (service *HTTPRestService) UpdateEndpointHelper(req cns.EndpointRequest) error {
+	if service.EndpointStateStore == nil {
+		return ErrStoreEmpty
+	}
+	logger.Printf("[UpdateEndpointState] Updating endpoint state for infra container %s", req.EndpointID)
+	if service.Options[common.OptManageEndpointState] == true {
+		if endpointInfo, ok := service.EndpointState[req.EndpointID]; ok {
+			logger.Printf("[UpdateEndpointState] Found existing endpoint state for infra container %s", req.EndpointID)
+			if req.HostVethName == "" && req.HnsEndpointID != "" {
```

Comment:
Validate the request is well-formed immediately, before anything else.

Reply:
addressed in the commit.